README.md
CrunchyXBMC
==========

Crunchyroll;xbmc
A XBMC plugin for Cruchyroll. 
This plugin a an amalgamation of some of my own code and other's patches.

Full source history at https://code.google.com/p/urlxl-repo/

Forum Posting: http://forum.xbmc.org/showthread.php?tid=129709

Disclaimer:
You will need a Crunchyroll account!
(No you don't have to pay for it. But it's cheap.)

Thanks to the following for code:

maruchan, Voinage, Zehd, mattrk, and le__


CHANGELOG.txt
[B]2.0.5[/B]
Added Auto-Resume of watched Videos
Increased Area to try to catch issues with loading save data
[B]2.0.3[/B]
More Clean Up
[B]2.0.1[/B]
Merge Frodo and Gotham Repos
Create Code to Push to XBMC Repo
[B]1.0/2.0[/B]
Clean up and final build for submission
[B]0.9 x12[/B]
Fixed EPs beyond 20 not showing
Removed POP link, since nothing in there played.
[B]0.9 x10[/B]
More Fixes
More Data
Changes to help non-US users login
Added anonymous user tracking
[B]Version 0.9 x2[/B]
More Fixes
Added Bigger Art for larger screens
[B]Version 0.9 x1[/B]
Lot's of Bug Fixes (le__)
[B]Version 0.9[/B]
Crunchyroll;xbmc beta1
Name change to reflect the major code changes
Based on the Crunchyroll plugin for Plex by mattrk
https://forums.plex.tv/index.php/topic/73626-rel-crunchyroll-plugin/
[B]Version 0.7.2.2[/B]
Added corrections for non-US / non-English sights.
[B]Version 0.7.2[/B]
Added ability to disable subtitles
Resolved long standing metadata bug
[B]Version 0.7.1[/B]
Added Workaround broken videos
[B]Version 0.7[/B]
RMTP is Back!
[B]Version 0.6.9.2[/B]
Added Portuguese (Brazil)
Removed Broken Seasons Link
[B]Version 0.6.9.0[/B]
BETA OF REBUILD
Disabled rtms streams, because they are broken again
[B]Version 0.6.5.2[/B]
Marked Plugin as BROKEN
[B]Version 0.6.5.1[/B]
Changed Subtitles Crypto for changes in XBMC Python in 12.2
Added Subtitle Language Selecton (Le__)
[B]Version 0.6.5[/B]
-Re-added alternative streaming method for playback on devices lacking rtmp playback
-Must be enabled in settings, disabled by default, SD playback only
[B]Version 0.6.0.5[/B]
-Resolve the plugin not showing on new Frodo installs.
-Force Refresh to show plugins
[B]Version 0.6.0.3[/B]
-Correct Playback issues for users of non-US Crunchyroll site. (le__)
-Corrected a subtitle decoding bug (mattrk)
-Props to le__ and  mattrk for their patches
[B]Version 0.6.0.2[/B]
-Roleback 0.6.0.1 due to broken Subtitle issue
[B]Version 0.6[/B]
-The OMG I Br0ked the Fix Fix
-Videos playback as before
-Thanks to Zehd for his help.
[B]Version 0.5[/B]
-The OMG Br0ken Vids Fix
+Known issue, this fix makes ALL playback SD.  -Sorry
[B]Version 0.4.2.1[/B]
-Fixed Edge a queue Case
[B]Version 0.4.2[/B]
-The Broken Cookie Fix
-To activate, open your Queue
[B]Version 0.4.1[/B]
-Fix for Series without Episode Names
[B]Version 0.4[/B]
-The Drama Update
+Known issue: Show listings max at 40 series
[B]Version 0.3.4.4[/B]
-Fix to Series Parcing caused by Null in Description field
-Thanks to dandel for this fix
[B]Version 0.3.4.3[/B]
-Fixes to norwegian language file
-Known Issue: non-ascii must be substituted
[B]Version 0.3.4.2[/B]
-Added login validataion for Queue
[B]Version 0.3.4.1[/B]
-Added language support
-now accepting language files
[B]Verison 0.3.4[/B]
-Updated code to display all videos in the correct format.
-Known Issue: 480p videos show as "SD" in XBMC; Log file shows correct resolution
-Please report videos that don't play in the correct resolution to google code.
[B]Verison 0.3.3.3[/B]
-Minor Change to handeling of shows without episodes
[B]Verison 0.3.3.2[/B]
-Updated Logo
[B]Version 0.3.3.1[/B]
-Minor Graphical Change to display of "Pre Release" shows
[B]Version 0.3.3[/B]
-Added Episode number and marking of "Pre Release" Episodes (Where Avalible)
[B]Version 0.3.2[/B]
-Corrected Playback of some 480p and 720p streams
-Known issue: Some streams may not play back due to the Video Encoding field not being picked up
[B]Version 0.3.1[/B]
-Updated Queue Pull to give notice of Slow steps
[B]Version 0.3[/B]
-Added Anime Simulcast Listing
[B]Version 0.2[/B]
-Added 1080p Playback (Where Avalible)
[B]Version 0.1[/B]
-First Build with Queue Functionality
-Known Issue: If CR is slow to login, Queue Plugin will crash
-Workaround: Play any video first
[B]Version 0.0[/B]
-Init. Fork
-Applied know patches to make plugin work with current Cunchyroll site
-Added Plugin.CryptoPy to Repo to make plugin installable on XBMC 11


default.py
"""
    CrunchyRoll;xbmc
    Copyright (C) 2012 - 2014 Matthew Beacher
    This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
"""
import sys
import xbmcaddon

#plugin constants
__plugin__ = "Crunchyrollxbmc"
__version__ = "2.0.2"
__XBMCBUILD__ = xbmc.getInfoLabel( "System.BuildVersion" )
__settings__ = xbmcaddon.Addon(id='plugin.video.crunchyroll-takeout')
__language__ = __settings__.getLocalizedString

print "[PLUGIN] '%s: version %s' initialized!" % (__plugin__, __version__)

if __name__ == "__main__":
    from resources.lib import crunchy_main
    crunchy_main.Main()
    
sys.modules.clear()


resources/lib/crunchy_main.py
# -*- coding: utf-8 -*-
"""
    CrunchyRoll;xbmc
    Copyright (C) 2012 - 2014 Matthew Beacher
    This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
"""
import sys
import os
import urllib
import xbmc
import xbmcgui
import xbmcplugin
import crunchy_json
__settings__ = sys.modules[ "__main__" ].__settings__




#####################################################################################

class updateArgs:

	def __init__(self, *args, **kwargs):
		for key, value in kwargs.iteritems():
			if value == 'None':
				kwargs[key] = None
			else:
				kwargs[key] = urllib.unquote_plus(kwargs[key])
		self.__dict__.update(kwargs)

class UI:
	
	def __init__(self):
		self.main = Main(checkMode = False)
		xbmcplugin.setContent(int(sys.argv[1]), 'movies')

	def endofdirectory(self, sortMethod = 'none'):
		# set sortmethod to something xbmc can use
		if sortMethod == 'title':
			sortMethod = xbmcplugin.SORT_METHOD_LABEL
		elif sortMethod == 'none':
			sortMethod = xbmcplugin.SORT_METHOD_NONE
		elif sortMethod == 'date':
			sortMethod = xbmcplugin.SORT_METHOD_DATE
		#Sort methods are required in library mode.
		xbmcplugin.addSortMethod(int(sys.argv[1]), sortMethod)
		#let xbmc know the script is done adding items to the list.
		dontAddToHierarchy = False
		xbmcplugin.endOfDirectory(handle = int(sys.argv[1]), updateListing = dontAddToHierarchy)
			
	def addItem(self, info, isFolder=True, total_items = 0):
		#Defaults in dict. Use 'None' instead of None so it is compatible for quote_plus in parseArgs
		info.setdefault('url','None')
		info.setdefault('Thumb','None')
		info.setdefault('Fanart_Image', xbmc.translatePath( __settings__.getAddonInfo('fanart') ) )
		info.setdefault('mode','None')
		info.setdefault('count','0')
		info.setdefault('filterx','None')
		info.setdefault('id','None')
		info.setdefault('series_id','None')
		info.setdefault('offset','0')
		info.setdefault('season','1')
		info.setdefault('series_id','0')
		info.setdefault('page_url','None')
		info.setdefault('complete','True')
		info.setdefault('showtype','None')
		info.setdefault('Title','None')
		info.setdefault('year','0')
		info.setdefault('plot','No description available.')
		#print info
		#create params for xbmcplugin module
		u = sys.argv[0]+\
			'?url='+urllib.quote_plus(info['url'])+\
			'&mode='+urllib.quote_plus(info['mode'])+\
			'&name='+urllib.quote_plus(info['Title'])+\
			'&id='+urllib.quote_plus(info['id'])+\
			'&count='+urllib.quote_plus(info['count'])+\
			'&series_id='+urllib.quote_plus(info['series_id'])+\
			'&filterx='+urllib.quote_plus(info['filterx'])+\
			'&offset='+urllib.quote_plus(info['offset'])+\
			'&icon='+urllib.quote_plus(info['Thumb'])+\
			'&complete='+urllib.quote_plus(info['complete'])+\
			'&fanart='+urllib.quote_plus(info['Fanart_Image'])+\
			'&season='+urllib.quote_plus(info['season'])+\
			'&showtype='+urllib.quote_plus(info['showtype'])+\
			'&year='+urllib.quote_plus(info['year'])+\
			'&plot='+urllib.quote_plus(info['plot'])
		#create list item
		#print u
		li=xbmcgui.ListItem(label = info['Title'], thumbnailImage = info['Thumb'])
		li.setInfo( type="Video", infoLabels={ "Title":info['Title'], "Plot":info['plot'], "Year":info['year']})
		li.setProperty( "Fanart_Image", info['Fanart_Image'])
		#for videos, replace context menu with queue and add to favorites
		if not isFolder:
			#li.setProperty("IsPlayable", "true")#let xbmc know this can be played, unlike a folder.
			#add context menu items to non-folder items.
			contextmenu = [('Queue Video', 'Action(Queue)')]
			li.addContextMenuItems( contextmenu )
		#for folders, completely remove contextmenu, as it is totally useless.
		else:
			li.addContextMenuItems([], replaceItems=True)
		#add item to list
		ok=xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=u, listitem=li, isFolder=isFolder, totalItems=total_items)

	def showMain(self):
                local_string = __settings__.getLocalizedString
                change_language = __settings__.getSetting("change_language")
                if crunchy_json.CrunchyJSON().loadShelf() is False:
                        self.addItem({'Title':'Session Failed: Check Login'})
                        self.endofdirectory()
                else:
                        if change_language != "0":
                                crunchy_json.CrunchyJSON().changeLocale()
                        Anime = local_string(35000).encode("utf8")
                        Drama = local_string(35004).encode("utf8")
                        Queue = local_string(35005).encode("utf8")
                        #Pop = local_string(35009).encode("utf8")
                        History = local_string(35011).encode("utf8")
                        self.addItem({'Title':Queue, 'mode':'queue'})
                        self.addItem({'Title':History, 'mode':'History'})
                        self.addItem({'Title':Anime, 'mode':'Channels','showtype':'Anime'})
                        self.addItem({'Title':Drama, 'mode':'Channels','showtype':'Drama'})
                        #self.addItem({'Title':Pop, 'mode':'Channels','showtype':'Pop'})
                        self.endofdirectory()

	def channels(self):
                local_string = __settings__.getLocalizedString
                popular = local_string(35003).encode("utf8")
                Simulcasts = local_string(35006).encode("utf8")
                Recently_Added = local_string(35002).encode("utf8")
                alpha = local_string(35012).encode("utf8")
                Browse_by_Genre = local_string(35007).encode("utf8")
                seasons = local_string(35010).encode("utf8")
                showtype = self.main.args.showtype
                self.addItem({'Title':popular,          'mode':'list_series', 'showtype':showtype, 'filterx':'popular', 'offset':'0'})
                self.addItem({'Title':Simulcasts,       'mode':'list_series', 'showtype':showtype, 'filterx':'simulcast', 'offset':'0'})
                self.addItem({'Title':Recently_Added,   'mode':'list_series', 'showtype':showtype, 'filterx':'updated', 'offset':'0'})
                self.addItem({'Title':alpha,            'mode':'list_series', 'showtype':showtype, 'filterx':'alpha', 'offset':'0'})
                self.addItem({'Title':Browse_by_Genre,  'mode':'list_categories', 'showtype':showtype, 'filterx':'genre', 'offset':'0'})
                self.addItem({'Title':seasons,          'mode':'list_categories', 'showtype':showtype, 'filterx':'season', 'offset':'0'})
                self.endofdirectory()
                
        def json_list_series(self):
                crunchy_json.CrunchyJSON().list_series(self.main.args.name, self.main.args.showtype, self.main.args.filterx, self.main.args.offset)

        def json_list_cat(self):
                crunchy_json.CrunchyJSON().list_categories(self.main.args.name, self.main.args.showtype, self.main.args.filterx)
                
	def json_list_collection(self):
		crunchy_json.CrunchyJSON().list_collections(self.main.args.series_id, self.main.args.name, self.main.args.count, self.main.args.icon, self.main.args.fanart)
		
	def json_list_media(self):
		crunchy_json.CrunchyJSON().list_media(self.main.args.id, self.main.args.filterx, self.main.args.count, self.main.args.complete, self.main.args.season, self.main.args.fanart)

	def json_History(self):
		crunchy_json.CrunchyJSON().History()

	def queue(self):
                crunchy_json.CrunchyJSON().Queue()
		
	def startVideo(self):
		crunchy_json.CrunchyJSON().startPlayback(self.main.args.name, self.main.args.url, self.main.args.id, self.main.args.icon)

	def Fail(self):
                local_string = __settings__.getLocalizedString
                badstuff = local_string(37007).encode("utf8")
		self.addItem({'Title':badstuff,'mode':'none'})
		print "Crunchyroll takeout --> crunchy_main.py checkMode fall through"
		self.endofdirectory()

class Main:

	def __init__(self, checkMode = True):
                crunchy_json.CrunchyJSON().loadShelf()
		self.parseArgs()
		if checkMode:
			self.checkMode()
                   
                                
	def parseArgs(self):
		# call updateArgs() with our formatted argv to create the self.args object
		if (sys.argv[2]):
			exec "self.args = updateArgs(%s')" % (sys.argv[2][1:].replace('&', "',").replace('=', "='"))
		else:
			# updateArgs will turn the 'None' into None.
			# Don't simply define it as None because unquote_plus in updateArgs will throw an exception.
			# This is a pretty ugly solution
			self.args = updateArgs(mode = 'None', url = 'None', name = 'None')

	def checkMode(self):
		mode = self.args.mode
		if mode is None:
			UI().showMain()
		elif mode == 'Channels':
			UI().channels()
		elif mode == 'list_series':
			UI().json_list_series()
		elif mode == 'list_categories':
			UI().json_list_cat()
		elif mode == 'list_coll':
			UI().json_list_collection()
		elif mode == 'list_media':
			UI().json_list_media()
		elif mode == 'History':
			UI().json_History()
		elif mode == 'queue':
			UI().queue()
                elif mode == 'videoplay':
			UI().startVideo()
                else:
                        UI().Fail()

		


